### PR TITLE
Add a framework and components to expose shared memory single copy to all components

### PR DESCRIPTION
**Pull request by hjelmn**
This PR moves the shared-memory single-copy code out of btl/sm and puts it in its own framework. This will allow other components to make use of this functionality without relying on btl/sm.

### CODE REVIEW

**Review by bwbarrett — DISMISSED** [review]
Overall, looks like a massive improvement.  A couple of nits included.

One larger question...  In the unlikely scenario that two processes come to different smsc selection answers, it looks like we're going to handle that poorly.  Maybe I'm reading the code wrong, but only one of the `get_endpoint()` calls seems to check the other side is actually using the same smsc.  Am I reading that wrong?

File: opal/mca/smsc/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# Copyright (c) 2010 Cisco Systems, Inc.  All rights reserved.
+# Copyright (c) 2021      Google, LLC. All rights reserved.
+# $COPYRIGHT$
+#
+# Additional copyrights may follow
+#
+# $HEADER$
+#
+
+AM_CPPFLAGS = $(LTDLINCL)
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/Makefile.am:21]: Done.

[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/Makefile.am:21]: Hmm. Got to remember this isn't piper. Will be fixed when i push the changes :)

File: opal/mca/smsc/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#                         Corporation.  All rights reserved.
+# Copyright (c) 2004-2005 The University of Tennessee and The University
+#                         of Tennessee Research Foundation.  All rights
+#                         reserved.
+# Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+#                         University of Stuttgart.  All rights reserved.
+# Copyright (c) 2004-2005 The Regents of the University of California.
+#                         All rights reserved.
+# Copyright (c) 2010 Cisco Systems, Inc.  All rights reserved.
+# Copyright (c) 2021      Google, LLC. All rights reserved.
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/Makefile.am:13]: Copy-and-paste from another Makefile.am. Fixing.

**Comment by gpaulsen** [comment]
We discussed this in the Web-ex Today.  Consensus was that it'd be good to get this into v5.0 as there are other optimizations and code-cleanup that Nathan's working on that this will enable.

File: opal/mca/smsc/Makefile.am (hunk 1 above)
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/Makefile.am]: Done.

File: opal/mca/smsc/Makefile.am (hunk 2 above)
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/Makefile.am:13]: Done.

File: opal/mca/smsc/base/smsc_base_impl.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+void mca_smsc_base_return_endpoint(mca_smsc_endpoint_t *endpoint)
+{
+    if (NULL == selected_module || NULL == endpoint) {
+        return;
+    }
+
+    selected_module->return_endpoint(selected_module, endpoint);
+}
+
+int mca_smsc_base_copy_to(mca_smsc_endpoint_t *endpoint, void *local_address, void *remote_address,
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/base/smsc_base_impl.c]: Done. Now handled in such a way to allow MCA direct (which was broken BTW-- another PR fixes it).

File: opal/mca/smsc/base/smsc_base_impl.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * $HEADER$
+ */
+
+#include "opal/mca/smsc/base/smsc_base_impl.h"
+
+mca_smsc_endpoint_t *mca_smsc_base_get_endpoint(opal_proc_t *peer_proc)
+{
+    if (NULL == selected_module) {
+        fprintf(stderr, "XXX No module\n");
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/base/smsc_base_impl.c]: Debug code. File removed. No longer relevant.

File: opal/runtime/opal_init.c [hunk 5]
```diff
@@ -619,11 +620,17 @@ int opal_init_util(int *pargc, char ***pargv)
  * versions of memcpy correctly configured.
  */
 static mca_base_framework_t *opal_init_frameworks[] = {
-    &opal_threads_base_framework,   &opal_hwloc_base_framework,
-    &opal_memcpy_base_framework,    &opal_memchecker_base_framework,
-    &opal_backtrace_base_framework, &opal_timer_base_framework,
-    &opal_shmem_base_framework,     &opal_reachable_base_framework,
-    &opal_pmix_base_framework,      NULL,
+    &opal_threads_base_framework,
+    &opal_hwloc_base_framework,
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/runtime/opal_init.c]: clang-format put them one per line. More readable that way but it does hide what changed. Reverted just the style change.

File: opal/mca/smsc/base/smsc_base_frame.c [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+        opal_output_verbose(MCA_BASE_VERBOSE_COMPONENT, opal_smsc_base_framework.framework_output,
+                            "mca_smsc_base_select: checking component %s",
+                            component->smsc_version.mca_component_name);
+
+        int ret = component->query();
+        if (OPAL_SUCCESS != ret) {
+            opal_output_verbose(MCA_BASE_VERBOSE_COMPONENT,
+                                opal_smsc_base_framework.framework_output,
+                                "mca_smsc_base_select: could not select component %s. enable "
```
[the remark this answers is not in the document]
Comment (hjelmn) [reply opal/mca/smsc/base/smsc_base_frame.c]: Done.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Review by jsquyres — CHANGES_REQUESTED** [review]
I get this warning when compiling:

```
2032 make[2]: Entering directory `/home/jsquyres/git/ompi/opal/mca/smsc'
2033   CC       base/smsc_base_frame.lo
2034 In file included from ../../../opal/mca/smsc/base/base.h:14,
2035                  from base/smsc_base_frame.c:27:
2036 ../../../opal/mca/smsc/smsc.h:216:5: warning: "MCA_opal_smsc_DIRECT_CALL" is not defined, evaluates to 0 [-Wundef]
2037  #if MCA_opal_smsc_DIRECT_CALL
2038      ^~~~~~~~~~~~~~~~~~~~~~~~~
```

(repeats for all the BTL SM + SMSC component source files, too)

Feels like that is significant...

**Comment by hjelmn** [comment]
@jsquyres Fixed. File missing from one of the commits. Please re-test.

**Review by jsquyres — APPROVED** [review]
Warning fixed.

**Pull request merged**